### PR TITLE
Autotools/vpath support

### LINKED ISSUE #10: Add autotools build system
In order for 3rd party projects that depend on this project to more easy build LevelDB using the makefiles, please add a autoconf compatible configure.ac.

### PR DESCRIPTION
**Pull request by kleetus**
closes #10
## Discussion

LevelDB currently has a hand-rolled Makefile and it seems to work well, especially for build systems that do not have a shell (Windows). What the build system for LevelDB does not have is support for Autotools, specifically leveraging autconf, autoreconf, and the macros that come along with that. 

LevelDB is, fundamentally, a project that should be a dependency to another project with its own build system. Many of these projects are also open-source and use Autotools. As such, those projects can easily integrate this project if, and only if, this project has at least a configure.ac file. Otherwise those projects need to hack in running "make -c" in their own Makefile.am's.
## This code adds
- a basic configure.ac and autogen.sh script
- a basic configure script
- a Makefile input file that mirrors the original Makefile, so that non-shell systems can still continue to use the Makefile

### CODE REVIEW

**Comment by laanwj** [comment]
Concept ACK

**Comment by da2ce7** [comment]
I don't think that the generated configure script should be included in the pull request.

Maybe update the .gitignore?

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,7 @@ Release
 Debug
 Benchmark
 vs2010.*
+autom4te.cache/
+config.log
+config.status
+!configure
```
Comment (laanwj) [inline .gitignore]: Agree that a generated script like configure shouldn't be in the repo. A typo, I think.

Comment (sipa) [inline .gitignore]: The pull request lists it as a feature, so I assume it's intentional.

**Comment by kleetus** [comment]
The addition of the configure script is intentional. The rationale is consistency with secp256k1 library which includes a configure shell script. Without the configure script, parent projects would need to run autoconf or autoreconf on leveldb.

**Comment by laanwj** [comment]
> Without the configure script, parent projects would need to run autoconf or autoreconf on leveldb.

The problem with committing build-generated scripts is that when developers do run `autoconf`, a repository file will have changes, which is easy to inadvertently commit, resulting in merge conflicts with other developers (as well as a lot of commit noise).

(as a convenience, ./configure script is part of the tarball created with `make dist`, but not the repository)

Also I suppose the only intended parent project here is bitcoin, which will handle calling the autoconf.

**Comment by laanwj** [comment]
Tested ACK apart from the configure nit - see https://github.com/bitcoin/bitcoin/issues/7466#issuecomment-182408931

**Comment by sipa** [comment]
There is no ./configure script included in the secp256k1 repository, by the way: https://github.com/bitcoin/secp256k1 nor https://github.com/bitcoin/bitcoin/tree/master/src/secp256k1

**Comment by kleetus** [comment]
> The problem with committing build-generated scripts is that when developers do run autoconf, a repository file will have changes, which is easy to inadvertently commit, resulting in merge conflicts with other developers (as well as a lot of commit noise).

Agreed! Let me fix this so that there is no configure script and that gitignore is properly aligned to ignore the output of autoreconf.

**Comment by laanwj** [comment]
@theuni can you take a look here please?

**Comment by theuni** [comment]
Hmm, I wonder if it would be simpler to just add our own simple Makefile.am here, rather than trying to pile on their stuff. I'll whip something up for comparison.

**Comment by laanwj** [comment]
Well this is already done, and checked it works.

**Comment by theuni** [comment]
Pushed an alternative to https://github.com/theuni/bitcoin/commit/3b714691956e623935929e5b174eceb8ab342bcd

The benefit is that it's directly integrated into our build without the need for another configure run. I'm not married to it, but I'm not a fan of adding more complexity to what should be a pretty straightforward build.

**Comment by laanwj** [comment]
I agree, it's a worthwhile alternative - however, please read the discussion in https://github.com/bitcoin/bitcoin/issues/7466 on how we got to this solution.

I think the advantage of your solution is that it can be maintained entirely outside of leveldb's tree (given that the .include is moved). Effectively it absorbs leveldb into bitcoin's build system.
On the other hand this solution would be better if leveldb upstream could be convinced to switch. It's more of a modular approach.

Not wedded to either though, I just want out of tree builds to work.

**Comment by laanwj** [comment]
@theuni I've tested your patch, but it fails on `test_bitcoin` and the bench.

Seems every executable now needs both `$(INT_LIBLEVELDB) $(INT_LIBMEMENV)` as well as `$(LIBLEVELDB) $(LIBMEMENV)` in its LDADD?

**Comment by theuni** [comment]
@laanwj I've pushed a cleaned up version here: https://github.com/theuni/bitcoin/tree/leveldb-integration. Having done that cleanly now, I do prefer it over the wrapper approach. That's not to diminish @kleetus's work, of course.

If you're ok with the approach, I'll PR it here.

**Comment by laanwj** [comment]
@theuni Looks good to me! You probably want to pr it at bitcoin/bitcoin though, not here.

**Comment by theuni** [comment]
Right, I forgot the file was moved up. PRing over there.

**Comment by laanwj** [comment]
Closing this in favor of the build system changes in Bitcoin Core.

It may still make sense to try to do this upstream.

**Pull request closed without merging**